supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/monitoring/CheckpointingConfiguration.scala

```diff
@@ -0,0 +1,74 @@
+package cromwell.backend.google.pipelines.common.monitoring
+
+import cromwell.backend.BackendJobDescriptor
+import cromwell.backend.io.WorkflowPaths
+import cromwell.core.path.Path
+
+import scala.concurrent.duration.FiniteDuration
+
+final class CheckpointingConfiguration(jobDescriptor: BackendJobDescriptor,
+                                       workflowPaths: WorkflowPaths,
+                                       commandDirectory: Path,
+                                       checkpointInterval: FiniteDuration
+                                      ) {
+  def checkpointFileCloud(checkpointFileName: String): String = {
+    // Fix the attempt at 1 because we always use the base directory to store the checkpoint file.
+    workflowPaths.toJobPaths(jobDescriptor.key.copy(attempt = 1), jobDescriptor.workflowDescriptor)
+      .callExecutionRoot.resolve("__checkpointing").resolve(checkpointFileName).toAbsolutePath.pathAsString
+  }
+  def tmpCheckpointFileCloud(checkpointFileName: String): String = checkpointFileCloud(checkpointFileName) + "-tmp"
+
+  def checkpointFileLocal(checkpointFileName: String): String = {
+    commandDirectory.resolve(checkpointFileName).toAbsolutePath.pathAsString
+  }
+  def tmpCheckpointFileLocal(checkpointFileName: String): String = checkpointFileLocal(checkpointFileName) + "-tmp"
+
+  def localizePreviousCheckpointCommand(checkpointFileName: String): String = {
+    val local = checkpointFileLocal(checkpointFileName)
+    val cloud = checkpointFileCloud(checkpointFileName)
+
+    s"gsutil cp $cloud $local || touch $local"
+  }
+
+  def checkpointingCommand(checkpointFilename: String, multilineActionSquasher: String => String): List[String] = {
+    val local = checkpointFileLocal(checkpointFilename)
+    val localTmp = tmpCheckpointFileLocal(checkpointFilename)
+    val cloud = checkpointFileCloud(checkpointFilename)
+    val cloudTmp = tmpCheckpointFileCloud(checkpointFilename)
+
+    val checkpointUploadScript =
+      s"""touch $local
+         |while true
+         |do
+         |  # Attempt to make a local copy of the checkpoint file
+         |  echo "CHECKPOINTING: Making local copy of $local"
+         |  COPY_SUCCESS="false"
+         |  while [ "$$COPY_SUCCESS" != "true" ]
+         |  do
+         |    PRE_COPY_TIMESTAMP="$$(stat -c'%Z' $local)"
+         |    cp $local $localTmp
+         |    if [ "$$PRE_COPY_TIMESTAMP" == "$$(stat -c'%Z' $local)" ]
+         |    then
+         |      COPY_SUCCESS="true"
+         |    else
+         |      echo "CHECKPOINTING: $local was modified while trying to make a local copy. Will retry in 10s..."
+         |      sleep 10
+         |    fi
+         |  done
+         |
+         |  # Perform the upload:
+         |  echo "CHECKPOINTING: Uploading new checkpoint content"
+         |  gsutil -m mv $localTmp $cloudTmp
+         |  echo "CHECKPOINTING: Replacing cloud checkpoint file with new content"
+         |  gsutil -m mv $cloudTmp $cloud
```

Comment:
I can see why we operate on a temp copy locally because the tool may be actively writing to it. But why do we need temp in the cloud backup location? Can we overwrite in place? Probably makes no difference, just curious.

Reply:
This was how the hackathon project did it, and it seemed sensible enough. I think the reasoning is that `gsutil cp` is not an atomic operation, and if the checkpoint upload is interrupted then the result in gcs is invalid. So instead, upload to a temporary location and then do an atomic `gsutil mv` to replace the previous content.